pkg/common-controller/groupsnapshot_finalizer_test.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common_controller
+
+import (
+	"testing"
+
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	"github.com/kubernetes-csi/external-snapshotter/v8/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+)
+
+func TestGroupSnapshotFinalizer(t *testing.T) {
+	tests := []controllerTest{
+		{
+			name: "7-1 - finalizer is added on dynamically provisioned group contents if the storage class specify a Deletion retain policy",
+			initialGroupSnapshots: newGroupSnapshotArray(
+				"snap-1-1", "snapuid1-1", map[string]string{
+					"app.kubernetes.io/name": "postgresql",
+				},
+				"", classGold, "groupsnapcontent-snapuid1-1", &False, nil, nil, false, false, nil,
+			),
+			expectedGroupSnapshots: withGroupSnapshotFinalizers(
+				newGroupSnapshotArray(
+					"snap-1-1", "snapuid1-1", map[string]string{
+						"app.kubernetes.io/name": "postgresql",
+					},
+					"", classGold, "groupsnapcontent-snapuid1-1", &False, nil, nil, false, false, nil,
+				),
+				utils.VolumeGroupSnapshotBoundFinalizer,
+			),
+			initialGroupContents: newGroupSnapshotContentArray(
+				"groupsnapcontent-snapuid1-1", "snapuid1-1", "snap-1-1", "snapshot-handle", classGold, []string{
+					"1-pv-handle6-1",
+					"2-pv-handle6-1",
+				}, "", deletionPolicy, nil, false, false,
+			),
+			expectedGroupContents: newGroupSnapshotContentArray(
+				"groupsnapcontent-snapuid1-1", "snapuid1-1", "snap-1-1", "snapshot-handle", classGold, []string{
+					"1-pv-handle6-1",
+					"2-pv-handle6-1",
+				}, "", deletionPolicy, nil, false, false,
+			),
+			initialClaims: withClaimLabels(
+				newClaimCoupleArray("claim1-1", "pvc-uid6-1", "1Gi", "volume6-1", v1.ClaimBound, &classGold),
+				map[string]string{
+					"app.kubernetes.io/name": "postgresql",
+				}),
+			initialVolumes: newVolumeCoupleArray("volume6-1", "pv-uid6-1", "pv-handle6-1", "1Gi", "pvc-uid6-1", "claim1-1", v1.VolumeBound, v1.PersistentVolumeReclaimDelete, classGold),
+			errors:         noerrors,
+			test:           testSyncGroupSnapshot,
+			expectSuccess:  true,
+		},
+		{
+			name: "7-2 - finalizer is not added on dynamically provisioned group contents if the storage class specify the Retain retain policy",
```

Comment:
storage class -> volume group snapshot class?

Reply:
Yes! Thank you!